core/leader/leader.go

```diff
@@ -50,10 +44,45 @@ func Run(db *sql.DB, addr string, lead func(context.Context)) {
 	}
 	log.Messagef(ctx, "Using leaderKey: %q", l.key)
 
-	update(ctx, l)
-	for range time.Tick(5 * time.Second) {
-		update(ctx, l)
+	var leadCtx context.Context
+	var cancel func()
+	for leader := range leadershipChanges(ctx, l) {
```

Comment:
Yeah I guess either way it depends on the leadershipChanges loop behaving correctly. There are two invariants it provides:

- each value is the opposite of the previous value
- the first value is true

My suggestion just moved us from depending on that first invariant to depending on the second one.

On second thought, I think it's ok to depend on either one of those invariants, but we should probably document them both in leadershipChanges. So IMO this code is fine as-is (with the addition of a brief comment).

The two-channel option also seems ok, but it has the possibility of deadlocking if either loop gets out of sync. An advantage of your original approach here is that it would be easy for this loop to enforce the invariant by also tracking its own bool variable, if we wanted to be that paranoid. (I don't think it's necessary, but I wouldn't object, either.)